src/components/Transactions.js

```diff
@@ -27,15 +28,38 @@ class Transactions extends React.Component {
       loaded: false,
       hasAfter: false,
       hasBefore: false,
+      queryParams: this.obtainQueryParams(),
     }
   }
 
   componentDidMount() {
-    this.getData(true, null, null, '');
+    this.getData(this.state.queryParams);
 
     WebSocketHandler.on('network', this.handleWebsocket);
   }
 
+  componentDidUpdate(prevProps, prevState) {
+    const queryParams = this.obtainQueryParams();
+
+    // Do we have new URL params?
+    if (JSON.stringify(this.state.queryParams) !== JSON.stringify(queryParams)) {
+      // Fetch new data, unless query params were cleared and we were already in the most recent page
+      if (queryParams.hash || this.state.hasBefore) {
+        this.getData(queryParams);
+      }
+      this.setState({ queryParams: queryParams });
+    }
+  }
+
+  obtainQueryParams() {
```

Comment:
In this method here we assume all parameters we expect will be on the URL. If they are not there, the page will get stuck loading forever.

We could add an error handling like:

```
if (!(params.has('ts')) || !(params.has('hash')) || !(params.has('page'))) {
  window.history.replaceState({}, "", window.location.pathname); // The same URL param clean you've used
  return { timestamp: null, hash: null, page: null }
}
```

Something like that. What do you think?

Reply:
The method has actually no such assumption. I could reproduce the "loading forever" issue you mentioned, by providing only `?hash=123`. However, this issue already exists and is more general: since the component currently [lacks error handling](https://github.com/HathorNetwork/hathor-explorer/blob/bb4e1998206835c64ac7957b7607d8e208b0b4f5/src/components/Transactions.js#L104-L107), any error on the initial API call would cause the same effect.

My understanding is that the above issue already exists and this change doesn't make it worse. If you agree, I propose we fix error handling in a separate PR.

Technical note: axiosInstance.js is not propagating rejections in its [`onRejected`](https://github.com/HathorNetwork/hathor-explorer/blob/bb4e1998206835c64ac7957b7607d8e208b0b4f5/src/api/axiosInstance.js#L22-L24) callback; it should `return Promise.reject(error)`.

On the other hand, errors on the subsequent API calls, although a bit less impactful, are made worse by this PR: since the URL gets updated upfront, there will be no effect clicking the same Next/Previous button again; the other button will force a new API call. Before this change, clicking the same button again would always trigger the API call. We should be able to fix that by updating `state.queryParams` only after we get a successful response, so that this state variable means "the latest queryParams we obtained results for". Thoughts?

Some context on `obtainQueryParams()`: it extracts the relevant URL params to a flat object format which is eventually stored in `state.queryParams`. It is used exactly twice:
1. On constructor to initialize state, which value is then used by `componentDidMount()` for the initial API call
2. On componentDidUpdate (also triggered on URL changes) to whether the params have been changed and if so, trigger a new API call.

Reply:
I agree that we were not handling error right now but having the query params in the URL is easier for the user to send a bad request.

However I've checked the API and it's not properly handling the expected parameters there. I think before any error handling in the client side I should add some in the API.

I will create an issue in the full node to improve this error handling on the API and another one to add error handling here in the client. Is that ok?

Reply:
Sure. Agreed!

I've pushed a new version with the following changes:
* Address all comments so far
* Improved URL handling to preserve other parameters and fragment identifies (aka hash) if they exist
* Update `state.queryParams` only after getting successful response, to mitigate issues as described above. Now if a server error occurs, the user can simply retry by clicking the pagination button again. As discussed, proper error handling, like displaying an error message, should be implemented in a separate PR.